### PR TITLE
Adding support for startup value for servos

### PR DESCRIPTION
**Pull request by pyjamasam**
This allows for a startup angle or pulse-width to be defined in the config file.  Keeps servos from moving around randomly when power is re-applied to them.

Presently only supports softpwm as the servo stuff is hard coded to use that.

Signed-off-by: Chris Whiteford <github@chrisandtennille.com>

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks.  Unfortunately, the changes to mcu.py are not correct - the config must be static and it's also not valid to extract toolhead timing during the config process.  Typically something like this would be done by catching the printer_state("ready") callback within the extras module.

Separately, when changing existing code, please keep within the style of that code - specifically, limit whitespace changes and wrap lines to 80 characters.

-Kevin

**Comment by pyjamasam** [comment]
Thanks for the feedback.  Hopefully the updated code is better aligned with how things should be and addresses your comments. 
This was my first foray into coding for klipper and I wasn't quite sure how to handle the startup stuff.

Let me know if anything else stands out as incorrect and I'll venture to get it fixed up.  
This feature keeps my purge wiper arm from continually creeping when I power off and on my printer.

chris.

File: klippy/extras/servo.py [hunk 1]
```diff
@@ -28,28 +28,51 @@ def __init__(self, config):
         self.gcode.register_mux_command("SET_SERVO", "SERVO", servo_name,
                                         self.cmd_SET_SERVO,
                                         desc=self.cmd_SET_SERVO_help)
-    def set_pwm(self, print_time, value):
+
+        self.initial_pwm_value = -1
```
Comment (KevinOConnor) [inline klippy/extras/servo.py]: Please use None instead of -1.

File: klippy/extras/servo.py [hunk 2]
```diff
@@ -28,28 +28,51 @@ def __init__(self, config):
...
+        #Check to see if an initial angle or pulse width is configured and
+        # set it as required
+        initial_angle = config.getfloat('initial_angle', -1)
```
Comment (KevinOConnor) [inline klippy/extras/servo.py]: This should have minval and maxval settings to ensure the config is sane.

File: klippy/extras/servo.py [hunk 3]
```diff
@@ -28,28 +28,51 @@ def __init__(self, config):
...
+        else:
+            initial_pulse_width = config.getfloat('initial_pulse_width', -1)
+            if initial_pulse_width != -1:
+                self.initial_pwm_value = round(self._get_pwm_from_pulse_width(
+                    initial_pulse_width),3)
+
+    def printer_state(self, state):
+        if state == 'ready':
+            print_time = self.printer.lookup_object('toolhead').get_last_move_time()
+            self._set_pwm(print_time, self.initial_pwm_value)
```
Comment (KevinOConnor) [inline klippy/extras/servo.py]: It's not clear to me if this is correct when an initial value is not set in the config.  I'd say only run set_pwm() if the user has specified an initial value.

File: klippy/extras/servo.py [hunk 4]
```diff
@@ -28,28 +28,51 @@ def __init__(self, config):
...
-        self.set_pwm(print_time, width * self.width_to_value)
-    def set_pulse_width(self, print_time, width):
+        return width * self.width_to_value
+
+    def _get_pwm_from_pulse_width(self, width):
         width = max(self.min_width, min(self.max_width, width))
-        self.set_pwm(print_time, width * self.width_to_value)
+        return width * self.width_to_value
+
+
```
Comment (KevinOConnor) [inline klippy/extras/servo.py]: It's best to not add whitespace changes (eg, adding blank lines) to a commit that also has functional changes.

**Comment by KevinOConnor** [comment]
Also, there should be an update to example-extras.cfg with documentation for the two new config settings for the servo code.

File: klippy/extras/servo.py (hunk 1 above)
Comment (pyjamasam) [reply klippy/extras/servo.py]: Fixed up in the latest commit

File: klippy/extras/servo.py (hunk 3 above)
Comment (pyjamasam) [reply klippy/extras/servo.py]: Fixed up in the latest commit

File: klippy/extras/servo.py (hunk 2 above)
Comment (pyjamasam) [reply klippy/extras/servo.py]: _get_pwm_from_angle does a min and max check (0 and max_angle) before sorting out the pwm value. (as does _get_pwm_from_pulse_width) so in either case the values entered in the config file are sanitized before being converted to pwm values.

**Comment by pyjamasam** [comment]
Requested changes made.

On your last point about raising a config error, currently the code only calls get_float for initial_pulse_width if get_float for initial_angle results in a None result.  The warnings in the config file were added because it wasn't obvious that initial_pulse_width was being ignored if initial_angle was set.

If there another way you'd like the error to be done?  Can you point me to an example.  I'd be glad to update if the code isn't what you had intended.

chris.

**Comment by KevinOConnor** [comment]
Thanks.

> On your last point about raising a config error, currently the code only calls get_float for initial_pulse_width if get_float for initial_angle results in a None result. The warnings in the config file were added because it wasn't obvious that initial_pulse_width was being ignored if initial_angle was set.

Okay, I missed that.  I guess I was confused by the wording in the config description.  Maybe adding the following to the description of both parameters would be more clear: "The default is to not set any value until the first SET_SERVO command." and add to initial_pulse_width: "This parameter is only valid if 'initial_angle' is not set."

-Kevin

**Pull request merged**